GVFS/GVFS.Common/NamedPipes/NamedPipeStreamReader.cs

```diff
@@ -0,0 +1,67 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+
+namespace GVFS.Common.NamedPipes
+{
+    /// <summary>
+    ///   The GVFS NamePipeProtocol:
+    ///       GVFS can send "messages" over the wire. A message is defined as a
+    ///       sequence of UTF8 encoded text whose end is delineated by a TerminatorByte.
+    /// </summary>
+    public class NamedPipeStreamReader
+    {
+        public const int BufferLength = 1024;
+
+        private const byte TerminatorByte = 0x3;
+
+        private byte[] buffer;
+        private int bufferContentLength;
+
+        private Stream stream;
+
+        public NamedPipeStreamReader(Stream stream)
+        {
+            this.stream = stream;
+            this.buffer = new byte[NamedPipeStreamReader.BufferLength];
+        }
+
+        public string ReadMessage()
+        {
+            List<byte> bytes = new List<byte>();
+
+            int bytesRead = this.Read();
+            if (bytesRead == 0)
+            {
+                return null;
+            }
+
+            while (true)
+            {
+                if (this.buffer[bytesRead - 1] == TerminatorByte)
+                {
+                    bytes.AddRange(new ArraySegment<byte>(this.buffer, 0, bytesRead - 1));
+                    break;
+                }
+
+                bytes.AddRange(new ArraySegment<byte>(this.buffer, 0, bytesRead));
+                bytesRead = this.Read();
+                Debug.Assert(bytesRead != 0, "Not expecting 0 bytes at this time");

```

Comment:
Please use exceptions rather than debug asserts

Reply:
How do we want to handle "unexpected" messages  - messages that do not end with the expected terminating byte? 

We could throw an exception - but then I suspect we also would want a caller to catch this exception and trace what happened. Alternatively, we could trace this occurance and return null...

Preferences?

Reply:
We already handle and log exceptions from all thread entry points, so throwing is equivalent to: Something went horribly wrong, log it and shut down.

If this situation ever happens, we've got a badly misbehaving client, or potentially mismatched versions installed on a box. In any case, it's extremely dangerous to keep going, so throwing seems like the best option.